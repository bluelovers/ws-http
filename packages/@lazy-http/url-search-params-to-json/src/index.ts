
export type IEntriesLike<K extends string, V extends unknown> = [key: K, val: V][] | IterableIterator<[K, V]>

export interface IOptions
{
	removeSquareBracketsArrayKey?: boolean,
	squareBracketsArrayKey?: boolean,
}

/**
 * @see https://stackoverflow.com/a/52539264/4563339
 */
export function entriesToJSON<T extends Record<string, any>>(entries: IEntriesLike<string, unknown>, options?: IOptions)
{
	const { removeSquareBracketsArrayKey = false, squareBracketsArrayKey = false } = options ?? {};

	const ks = new Set<string>();

	let json : T = {} as any;
	for (const [key, val] of entries)
	{
		if (squareBracketsArrayKey && key.endsWith('[]'))
		{
			// @ts-ignore
			json[key] = [];
		}

		if (Object.prototype.hasOwnProperty.call(json, key))
		{
			// if the current key is already an array, we'll add the value to it
			if (Array.isArray(json[key]))
			{
				json[key].push(val)
			}
			else
			{
				// if it's not an array, but contains a value, we'll convert it into an array
				// and add the current value to it
				// @ts-ignore
				json[key] = [json[key], val];
			}

			ks.add(key)
		}
		else
		{
			// plain assignment if no special case is present
			// @ts-ignore
			json[key] = val;
		}
	}

	if (removeSquareBracketsArrayKey)
	{
		json = removeSquareBracketsFromJSON(json, ks)
	}

	return json
}

export function removeSquareBracketsFromJSON<T extends Record<string, unknown>>(json: T, ks?: Set<string>)
{
	ks ??= new Set(Object.keys(json));

	json = {
		...json,
	}

	ks.forEach(key =>
	{
		const val = json[key];
		if (key.endsWith('[]') && Array.isArray(val))
		{
			const key2 = key.replace(/\[\]$/, '');
			if (!key2.length)
			{
				throw new TypeError(`Invalid key: '${key}'`)
			}
			else if (Object.prototype.hasOwnProperty.call(json, key2))
			{
				throw new TypeError(`'${key2}' already exists in keys`)
			}
			delete json[key]
			// @ts-ignore
			json[key2] = val;
		}
	})

	return json
}

/**
 * @see https://stackoverflow.com/a/52539264/4563339
 */
export function searchParamsToJSON<T extends Record<string, any>>(sp: URLSearchParams, options?: IOptions)
{
	return entriesToJSON<T>(sp.entries(), options)
}

export function searchParamsStringToJSON<T extends Record<string, any>>(sp: string, options?: IOptions)
{
	return searchParamsToJSON<T>(new URLSearchParams(sp), options)
}

export function jsonToEntries(json: Record<string, unknown>, options?: IOptions)
{
	const { squareBracketsArrayKey = false, removeSquareBracketsArrayKey = false } = options ?? {};

	if (removeSquareBracketsArrayKey)
	{
		json = removeSquareBracketsFromJSON(json)
	}

	return Object.entries(json)
		.reduce((entries, [key, val]) => {
			if (Array.isArray(val))
			{
				let bool = key.endsWith('[]');

				if (squareBracketsArrayKey && !bool)
				{
					key += '[]';
				}

				val.forEach(v => {
					entries.push([key, v])
				})
			}
			else
			{
				entries.push([key, val])
			}

			return entries
		}, [] as [string, unknown][])
}

export function jsonToSearchParams(json: Record<string, unknown>, options?: IOptions)
{
	return new URLSearchParams(jsonToEntries(json, options) as any)
}

export default searchParamsToJSON
