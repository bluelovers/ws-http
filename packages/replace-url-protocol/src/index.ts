import { notStrictEqual, strictEqual } from 'assert';

export function replaceProtocol(href: string, protocol: string)
{
	return href.replace(/^[^:]+:/, protocol)
}

/**
 * @internal
 */
export function _fixReplaceURLProtocol<T extends URL, P extends string = string>(url: T, oldProtocol: string, newProtocol: P): asserts url is T & {
	protocol: P,
}
{
	if (_isSameProtocol(url.protocol, oldProtocol))
	{
		url.href = replaceProtocol(url.href, newProtocol);
		assertProtocolNotEqual(url.protocol, oldProtocol);
	}
}

/**
 * @internal
 */
export function _isSameProtocol<T extends string>(actualProtocol: string, expectedProtocol: T): actualProtocol is T
{
	return actualProtocol === expectedProtocol
}

export function assertProtocolNotEqual(actualProtocol: string, expectedProtocol: string)
{
	notStrictEqual(actualProtocol, expectedProtocol);
}

export function assertProtocolEqual<T extends string>(actualProtocol: string, expectedProtocol: T): asserts actualProtocol is T
{
	strictEqual(actualProtocol, expectedProtocol);
}

export function replaceThisProtocol<T extends URL>(this: T, protocol: string)
{
	return replaceURLProtocol(this, protocol);
}

/**
 * helper for avoid node.js can't update protocol for some url
 *
 * @see https://github.com/nodejs/node/issues/39732
 */
export function replaceURLProtocol<T extends URL, P extends string = string>(url: T, protocol: P)
{
	const old = url.protocol;

	if (!_isSameProtocol(old, protocol))
	{
		url.protocol = protocol
		_fixReplaceURLProtocol(url, old, protocol);
	}

	return url;
}

export default replaceURLProtocol
