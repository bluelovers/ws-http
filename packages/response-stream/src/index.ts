/**
 * Created by user on 2020/2/26.
 */

import { PassThrough } from "stream";
import { ServerResponse } from "http";

export function responseStream(serverResponse: ServerResponse, data)
{
	let readStream = new PassThrough();
	readStream.end(data);
	return readStream.pipe(serverResponse);
}

export default responseStream
