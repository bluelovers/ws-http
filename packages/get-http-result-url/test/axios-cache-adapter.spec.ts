import { basename, extname } from 'path';
import { resultToURL } from '../src/index';

describe(basename(__filename, extname(__filename)), () =>
{

	test(`fromCache`, () =>
	{

		let actual = resultToURL({
			config: {
				"url": "https://www.novelstar.com.tw/books/",
				"method": "get",
				"headers": {
					"Accept": "application/json, text/plain, */*",
					"Referer": "https://www.novelstar.com.tw/",
				},
				"params": {
					"p": 1,
				},
				"transformRequest": [
					null,
				],
				"transformResponse": [
					null,
				],
				"timeout": 0,
				"withCredentials": true,
				"responseType": "arraybuffer",
				"xsrfCookieName": "XSRF-TOKEN",
				"xsrfHeaderName": "X-XSRF-TOKEN",
				"maxContentLength": -1,
				"maxBodyLength": -1,
				"cache": {
					"maxAge": 21600000,
				},
				"raxConfig": {
					"retry": 1,
					"retryDelay": 1000,
				},
				"maxRedirects": 0,
				"baseURL": "https://www.novelstar.com.tw/",
			},
			request: {
				"fromCache": true,
			},
		});
		let expected = 'https://www.novelstar.com.tw/books/?p=1';

		expect(actual).toHaveProperty('href', expected);
		expect(actual).toMatchSnapshot();

	});

})
