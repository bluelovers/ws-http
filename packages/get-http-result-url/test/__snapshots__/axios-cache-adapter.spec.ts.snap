// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`axios-cache-adapter.spec fromCache 1`] = `"https://www.novelstar.com.tw/books/?p=1"`;
